A canary roadmap (#58)
strimzi/proposals: 037-kafka-canary-roadmap.md

# Kafka Canary roadmap

This proposal describes a possible future set of improvement and features for the [Strimzi Canary](https://github.com/strimzi/strimzi-canary)

## Current situation

The project [README.md](https://github.com/strimzi/strimzi-canary/blob/main/README.md) explains the current feature set.

The Canary was added to Strimzi under [proposal 13](013-kafka-canary.md).

> The implementation is in golang and uses the [Sarama](https://github.com/Shopify/sarama/) Kafka client.

Reviewer: Just a nit ... here and in other places ... I think the usual use is Golang with capital `G`

## Motivation

We believe the Canary component could deliver significant value for Strimzi users, for example at the basis for an SLI on cluster availability and health.
However, these Canary improvements are held back by technical and non-technical issues:

* We've experienced [numerous bugs]((https://github.com/strimzi/strimzi-canary/issues?q=is%3Aissue+is%3Aclosed)) in the Kafka client library being used, and believe we're likely to see a long tail of further bugs.

> * The client library does not support all the features we want. Those features would needed to realize the Canary's true potential.

Reviewer 2: ```suggestion
* The client library does not support all the features we want. Those features would be needed to realize the Canary's true potential.
```

So over all we expect our community would be better able to engage with the Canary component if it was implemented in Java.

### Desirable features

There are several client features we would like to be able use.

* Support for SASL `OAUTHBEARER` mechanism. Sarama supports this.
* Support for forcing leader elections. Sarama doesn't support this.
* Support for transactional produce. Sarama doesn't seem to support this, at least not in a usable way. You can send the transaction-related requests, but there is no client library support akin to the `TransactionManager`, for ensuring the RPCs are sent correctly. 

### Differential observations

By running multiple producers and/or consumers, a canary could more accurately indicate where problems lay. 

For example, in a cluster with multiple listeners using different authentication mechanisms, if produce requests via a `SASL/OAUTHBEARER`-authenticated listener fail, but succeed via the `SASL/PLAIN`-authenticated listener you can infer that the problem is with the login mechanism. 

There are many dimensions in which you can do this: 

* DNS resolution

> * TLS (e.g. different versions or detecting imminent broker certificate expiry

Reviewer 2: ```suggestion
* TLS (e.g. different versions or detecting imminent broker certificate expiry)
```

* Producing to replicas on different volumes

> * With/without transactional producers, to flush out problems with transaction coordinators.

Reviewer 2: This is would be a really good thing to have and something that currently (AFAIK) we can't do with the samara client.

* Similarly with/without using consumer groups. Using different group ids to detect problems with group coordinators.

### Integration examples

Currently the Canary provides a binary and nothing more. 
This limits uptake of the canary because to use it for its intended purpose you need to configure prometheus scraping, alerting and dashboards to make sense of what the canary is telling you. 

> We would like to provide example of all these things, as we [do for the operator](https://github.com/strimzi/strimzi-kafka-operator/tree/main/examples/metrics).

Reviewer 3: We have this kind of stuff here https://github.com/strimzi/strimzi-canary/tree/main/packaging/examples/metrics
It provides the `PodMonitor` for configuring Prometheus and a Grafana dashboard example.

Author: Thanks, I wasn't aware of that.

### Actionable documentation

> Assuming the differential observations feature proves to provide worthwhile information, we would proceed to also provide documented procedures recommending corrective actions based on the specific faults it reported.

Reviewer 2: In the long term, some of these actions could be carried out automatically by Strimzi itself (assuming it is aware of a canary running).

## Affected/not affected projects

This affects the Canary project only.

## Compatibility

The API of the existing canary comprises:

* The configuration env vars
* Metrics (via the Prometheus scrape endpoint)
* The status and health endpoints

### Configuration
While we can strive to minimize incompatibilities, it seems of questionable value.
The Java clients are normally configured via `.properties` files.
Providing a compatibility mapping from env var to properties would be a lot of work, given that the canary hasn't seem very wide usage.
Some options would become obselete (e.g. `KAFKA_VERSION`), others are incompatible with how the Java clients work (e.g. `KAFKA_BOOTSTRAP_BACKOFF_MAX_ATTEMPTS`).

> So it seems better to provide a clean break.

Reviewer 2: Agreed.

So generally speaking, we would expect the Canary to experience fewer issues due to client bugs.
It would also necessarily support the latest versions of the Kafka protocol, which would enable support for more features.

The Canary also stands out as being the only Strimzi project that's written in Golang.
There's nothing wrong with Golang, but there are non-technical drawbacks to this situation.
Most existing Strimzi developers have more experience with Java than Golang.
Some don't really know Golang at all.
It's difficult to motivate people to use another language if they're using it only for one relatively small part of their job.
Likewise, existing Golang developers are not going to be drawn to contributing to Strimzi on account of a single Golang component.
So over all we expect our community would be better able to engage with the Canary component if it was implemented in Java.

### Desirable features

There are several client features we would like to be able use.

* Support for SASL `OAUTHBEARER` mechanism. Sarama supports this.
* Support for forcing leader elections. Sarama doesn't support this.

> * Support for transactional produce. Sarama doesn't seem to support this, at least not in a usable way. You can send the transaction-related requests, but there is no client library support akin to the `TransactionManager`, for ensuring the RPCs are sent correctly. 

Reviewer 3: it's not clear (at least to me) if "desirable features" are something for the future or strictly related to this proposal so the first implementation of it. Can you clarify it please?

### Differential observations

By running multiple producers and/or consumers, a canary could more accurately indicate where problems lay. 

For example, in a cluster with multiple listeners using different authentication mechanisms, if produce requests via a `SASL/OAUTHBEARER`-authenticated listener fail, but succeed via the `SASL/PLAIN`-authenticated listener you can infer that the problem is with the login mechanism. 

There are many dimensions in which you can do this: 

* DNS resolution
* TLS (e.g. different versions or detecting imminent broker certificate expiry)
* Producing to replicas on different volumes
* With/without transactional producers, to flush out problems with transaction coordinators.
* Similarly with/without using consumer groups. Using different group ids to detect problems with group coordinators.

### Actionable documentation

Assuming the differential observations feature proves to provide worthwhile information, we would proceed to also provide documented procedures recommending corrective actions based on the specific faults it reported.

### Changes and removals

> * Logging mechanism will be changed to Log4j2, to have it consistent with other Strimzi components

Reviewer 3: "will use" ... it's clear it "will be changed" because we are moving from Golang to Java :-)

  * The log levels will have to be changed from `0, 1, 2` to `INFO, DEBUG, WARN, TRACE`.
* `SARAMA_LOG_ENABLED` environment variable will be removed

> * `VERBOSITY_LOG_LEVEL` will be changed to `CANARY_LOG_LEVEL` and used inside the `log4j2.properties` file - default will be `INFO`

Reviewer 3: clarify it's related to the first bullet otherwise they seem to be two different things.

* Implementation of dynamic watcher will be removed 

>   * For logging changes the dynamic reconfiguration will be implemented (similarly to operators repository)

Reviewer 3: provided by log4j2? clarify ...

## Affected/not affected projects

This affects the Canary project only.

## Compatibility

The API of the existing canary comprises:

* The configuration env vars
* Metrics (via the Prometheus scrape endpoint)
* The status and health endpoints

### Configuration
While we can strive to minimize incompatibilities, it seems of questionable value.
The Java clients are normally configured via `.properties` files.
Providing a compatibility mapping from env var to properties would be a lot of work, given that the canary hasn't seem very wide usage.
Some options would become obselete (e.g. `KAFKA_VERSION`), others are incompatible with how the Java clients work (e.g. `KAFKA_BOOTSTRAP_BACKOFF_MAX_ATTEMPTS`).
So it seems better to provide a clean break.

### Metrics

> It seems likely that metric names and labels can be maintained, so in that respect we could be compatible. 

Reviewer 3: Link to the canary Golang metrics list would help https://github.com/strimzi/strimzi-canary#metrics-1
Are we taking all the metrics?

Detailed assessment of whether the metric values were being measured in a compatible way is beyond the scope of this document.
If measurements were incompatible it would require existing users to adjust thresholds used in any alerting they might have that was based on the old metrics.

### Status and Health endpoints

> Providing compatibility for these should be trivial.

Reviewer 3: Maybe a link to the section at the canary Golang implementation could help people to understand what you are talking about https://github.com/strimzi/strimzi-canary#endpoints

## Rejected alternatives

The main alternative would be to stick with Golang and Sarama.

> Doing this means:
> * the feature improvements be blocked, waiting for support to be added for the missing features.
> * the non-technical issues described would remain unaddressed. 

Reviewer 4: ```suggestion
However, by not switching to a Java implementation the issues highlighted in this proposal are unresolved:
* Feature improvements are blocked, waiting for support to be added for the missing features.
* Non-technical issues remain unaddressed. 
```

# Kafka Canary roadmap

> This proposal describes a possible future set of improvements and features for the [Strimzi Canary](https://github.com/strimzi/strimzi-canary) tool, which is used to check that Kafka clusters are operating correctly.  

Reviewer: It should not be a _possible future set_. You are proposing it, so the assumption should be that if it is approved it is done. If you are not sure about it (but I don't think so), then it needs to be reevaluated.

## Current situation

The Canary project [README.md](https://github.com/strimzi/strimzi-canary/blob/main/README.md) explains the current feature set.

The Canary was added to Strimzi under [proposal 13](013-kafka-canary.md).
The implementation is in Golang and uses the [Sarama](https://github.com/Shopify/sarama/) Kafka client.

## Motivation

> We believe the Canary component could deliver significant value for Strimzi users. For example, Canary could be used as a basis for an SLI on cluster availability and health.

Reviewer: I think the Canary is absolutely failing in this aspect as the information it provides is absolutely not user friendly and requires a lot of expertise to understand. This statement is a bit confusing for me -> it is not clear if it implies any changes (in that case they need to be elaborated on) or just repeats what was said before.

However, these Canary improvements are held back by the following technical and non-technical issues:

* We've experienced [numerous bugs]((https://github.com/strimzi/strimzi-canary/issues?q=is%3Aissue+is%3Aclosed)) in the Kafka client library being used, and believe we're likely to see a long tail of further bugs.

> * The client library does not support all the features we want. Those features would be needed to realize the Canary's true potential.

Reviewer: What features does it not support? You should provide a list.

* The client library is not (currently) evolving at the same pace as the Java clients, i.e. it seems to be falling further behind in terms of features.

> * The Strimzi community's lack of Golang expertise (or indeed enthusiasm) means it is somewhat unloved.

Reviewer: I would leave enthusiasm out of it.

## Proposal

### Re-implement in Java

The Kafka project itself only provides support for JVM clients.
Naturally those clients are extremely well-tested, with both current and older broker versions. 
So generally speaking, we would expect the Canary to experience fewer issues due to client bugs.
It would also necessarily support the latest versions of the Kafka protocol, which would enable support for more features.

The Canary also stands out as being the only Strimzi project that's written in Golang.
There's nothing wrong with Golang, but there are non-technical drawbacks to this situation.
Most existing Strimzi developers have more experience with Java than Golang.
Some don't really know Golang at all.
It's difficult to motivate people to use another language if they're using it only for one relatively small part of their job.
Likewise, existing Golang developers are not going to be drawn to contributing to Strimzi on account of a single Golang component.
So overall we expect our community would be better able to engage with the Canary component if it was implemented in Java.

### Desirable features

There are several client features we would like to be able use.

> * Support for SASL `OAUTHBEARER` mechanism. Sarama supports this.

Reviewer: I'm not really sure why is it listed here if Sarama supports it,

* Support for forcing leader elections. Sarama doesn't support this.
* Support for transactional produce. Sarama doesn't seem to support this, at least not in a usable way. You can send the transaction-related requests, but there is no client library support akin to the `TransactionManager`, for ensuring the RPCs are sent correctly. 

All of the desirable features will be implemented after Canary re-implementation.

### Differential observations

> By running multiple producers and consumers, Canary could more accurately indicate where problems lay. 

Reviewer: This seems to suggest there will be multiple producers and consumers used. But it seems to come out of the blue. It looks like this is missing some previous explanation?

> For example, in a cluster with multiple listeners using different authentication mechanisms, if produce requests fail using an `SASL/OAUTHBEARER`-authenticated listener but succeed using an`SASL/PLAIN`-authenticated listener, you can infer that the problem is with the login mechanism. 

Reviewer: If this is planned, you should probably explain more of how it will work and how will it be configured.

There are many dimensions in which you can run multiple producers and consumers: 

* DNS resolution

> * TLS (e.g. different versions or detecting imminent broker certificate expiry)

Reviewer: > detecting imminent broker certificate expiry

Not sure how would the Canary see it. It is IMHO also not its job.

> * Producing to replicas on different volumes

Reviewer: How does this relates to number of producers? Isn't this purely tied to placement of replicas?

* With/without transactional producers, to flush out problems with transaction coordinators.
* Similarly with/without using consumer groups. Using different group ids to detect problems with group coordinators.

> 

Reviewer: Double empty line not needed?

### Actionable documentation

> Assuming the differential observations feature proves to provide worthwhile information, we would proceed to also provide documented procedures recommending corrective actions based on the specific faults it reported.

Reviewer: Again. You are proposing this. So you should be sure about it?

  * The log levels will have to be changed from `0, 1, 2` to `INFO, DEBUG, WARN, TRACE`.
  * `VERBOSITY_LOG_LEVEL` will be changed to `CANARY_LOG_LEVEL` and used inside the `log4j2.properties` file - default will be `INFO`
* `SARAMA_LOG_ENABLED` environment variable will be removed
* Implementation of dynamic watcher will be removed 
  * For logging changes the dynamic reconfiguration will be used from Log4j2 (similarly to operators repository)

## Affected/not affected projects

This affects the Canary project only.

## Compatibility

The API of the existing Canary comprises the following:

* The configuration env vars
* Metrics (via the Prometheus scrape endpoint)
* The status and health endpoints

### Configuration

Java clients are normally configured via `.properties` files.
Providing a compatibility mapping from env var to properties would be a lot of work, given that the Canary hasn't seen very wide usage.
Some options would become obselete (e.g. `KAFKA_VERSION`), others are incompatible with how the Java clients work (e.g. `KAFKA_BOOTSTRAP_BACKOFF_MAX_ATTEMPTS`).
While we can strive to minimize incompatibilities, it seems of questionable value.

> So it seems better to provide a clean break.

Reviewer: This is fine. But you should also explain how will the configuration look like (perhps not here but in a separate part).

### Metrics

> It seems likely that most of the [metric names](https://github.com/strimzi/strimzi-canary#metrics-1) and labels can be maintained.

Reviewer: Why would they but not maintainable?

There are metrics which will be deleted or changed:

* Metrics that will be removed:
  * `topic_alter_assignments_error_total` - Canary in Java will not have dynamic reassignment
  * `consumer_timeout_join_group_total` - connected to Sarama client, Java clients don't have such error

* Metrics that will be changed:
  * `topic_describe_cluster_error_total` - will be changed to `describe_cluster_error_total`

* Metrics that need to be re-evaluated:
  * `producer_refresh_metadata_error_total` and `consumer_refresh_metadata_error_total`
    * with Sarama, there is a possibility to manually refresh metadata of both producer and consumer
    * there is no such possibility in Java

Detailed assessment of whether the metric values were being measured in a compatible way is beyond the scope of this document.
If measurements were incompatible, it would require existing users to adjust the thresholds used in any alerting they might have that was based on the old metrics.

### Status and Health endpoints

> Providing compatibility for [these](https://github.com/strimzi/strimzi-canary#endpoints) should be trivial.

Reviewer: You changed so many things in a backwards incompatible manner that this might not matter anymore.
